### DISCUSSION TITLE
My rule for lint: Prefer immutable collections

### DISCUSSION THREAD

**Discussion by FuryOfFools** [rrousselGit/riverpod#2930] [Q&A] [answered]
I've never written custom rules for a linter. Tell me how this could be done.
In general, the problem sounds something like this - "I'm using fast_immutable_collections, and I want to make a rule that says it's preferable to use `IList` instead of `List` for provider parameters and do the same with other collections".

**Comment by rrousselGit** [comment, marked as answer]
You'd use [custom_lint](pub.dev/packages/custom_lint). Follow its installation steps to create a package.

Then for the lint, you'd do `context.registry.addArgumentList`. This gives you a callback executed whenever there's an argument declared somewhere, with its type and more. You could then check the type to see if it's a List. And if so, warn.

**Reply by FuryOfFools** [reply]
If I understand correctly, this will work for all methods, not just providers? Is there any way to do this specifically for providers?

**Reply by rrousselGit** [reply]
You can detect that the class is annotated with @riverpod
